File: Data Structures & Algorithms/valid-palindrome-ii/submission-2.js
class Solution {
    /**
     * @param {string} s
     * @return {boolean}
     */
    validPalindrome(s) {
        let l = 0;
        let r = s.length -1;
  

        while(l < r){
            if(s[l] !== s[r] ){
                return (this.isPalindrome(s,l+1 ,r ) || this.isPalindrome(s, l, r-1))
            }
            l++;
            r--;
        }
        return true;


    }

    isPalindrome(s, l, r){
        let left = l;
        let right = r;
        while(left < right){
            if(s[left]!== s[right]){
                return false;
            }
            left++;
            right--;
        }
        return true;
    }
}

File: Data Structures & Algorithms/boats-to-save-people/submission-2.rs
impl Solution {
    pub fn num_rescue_boats(mut people: Vec<i32>, limit: i32) -> i32 {
        people.sort_unstable(); // Faster in-place sort for primitives
        
        let mut left: usize = 0;
        let mut right: usize = people.len();
        let mut boats = 0;

        while left < right {
            right -= 1; // Take the heaviest remaining person
            
            // Check if the lightest person can share this boat
            if left < right && people[left] + people[right] <= limit {
                left += 1;
            }
            
            boats += 1;
        }

        boats

    }
}
